Lowercase processor plugin

philipnrmn opened the pull request:
This PR adds a processor plugin which casts all uppercase metric fields to lowercase. Optionally, the original field may be preserved (ie both `Some_Metric` and `some_metric` would be sent). 

- Resolves [DCOS-43639](https://jira.mesosphere.com/browse/DCOS-43639)

philipnrmn committed:
Add process plugin to enforce lowercase metrics

branden reviewed (dismissed):
This looks great. Separating this logic into its own processor plugin is a good idea, I like its interface, and its tests are thorough. 👍

philipnrmn commented:
Thanks for the approvals! I've just realised that while lowercasing fields is all very well (and will work well for almost all telegraf plugins) the statsd plugin sends each statsd metric as a single influxdb-format metric. So we need to lowercase the metric name as well.

philipnrmn committed:
Cast names as well as fields to lowercase

philipnrmn commented:
I pushed a fresh commit, if you wouldn't mind taking another look?

gracedo reviewed (dismissed):
👍

janisz commented on plugins/processors/lowercase/lowercase.go:
```diff
@@ -0,0 +1,54 @@
+		for key, value := range metric.Fields() {
+			if strings.ContainsAny(key, capitals) {
+				metric.RemoveField(key)
+				metric.AddField(strings.ToLower(key), value)
```
I was thinking about the following simplification, but `metric.Copy()` does perform a shallow copy. I think we should add a comment here to describe why we remove and add a field instead of just alter its key.

```golang 
func (l *Lowercase) Apply(in ...telegraf.Metric) []telegraf.Metric {
	out := make([]telegraf.Metric, 0, len(in))
	for _, m := range in {
		if l.SendOriginal {
			out = append(out, m.Copy())
		}
		m.SetName(strings.ToLower(m.Name()))
		for _, field := range m.FieldList() {
			field.Key = strings.ToLower(field.Key)
		}
		out = append(out, m)
	}
	return out
}
```

janisz commented on plugins/processors/lowercase/lowercase.go:
```diff
@@ -0,0 +1,54 @@
+				metric.AddField(strings.ToLower(key), value)
+			}
+		}
+		if strings.ContainsAny(metric.Name(), capitals) {
```
Do we need this condition?

janisz commented on plugins/processors/lowercase/lowercase.go:
```diff
@@ -0,0 +1,54 @@
+			out = append(out, metric.Copy())
+		}
+		for key, value := range metric.Fields() {
+			if strings.ContainsAny(key, capitals) {
```
Do we need this condition?

janisz reviewed (dismissed):
LGTM only small comments.

janisz commented on plugins/processors/lowercase/lowercase.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (l *Lowercase) Apply(in ...telegraf.Metric) []telegraf.Metric {
+	out := []telegraf.Metric{}
```
Microptimisation but we know then lenght of this slice
`out := make([]telegraf.Metric, 0, len(in))`

philipnrmn replied:
We do if the `SendOriginals` flag is `false`. Otherwise we know it's not less than `len(in)`. I'll make the optimisation :+1:

janisz commented on plugins/processors/lowercase/lowercase.go:
```diff
@@ -0,0 +1,54 @@
+  # send_original = false
+`
+
+func (l *Lowercase) SampleConfig() string {
```
I think all functions should not be a pointer receiver. They do not modify the state of the structure.

philipnrmn replied:
It's true, they don't need to be pointer receivers  - but we have to implement them in order to satisfy the interface: https://github.com/dcos/telegraf/blob/1.7.2-dcos/processor.go

philipnrmn committed:
Test that lowercase metrics are not duplicated

With SendOriginals enabled, metrics with names or fields which contain uppercase
characters should be duplicated. However, metrics which do not should not be.

philipnrmn committed:
Benchmark testing for uppercase chars

philipnrmn committed:
Factor out logic in Apply

philipnrmn committed:
Optimise the out slice

philipnrmn commented:
@janisz I responded to all your feedback! thanks for the careful review.
